crates/ruff_python_formatter/src/expression/expr_slice.rs

```diff
@@ -1,23 +1,257 @@
+use crate::comments::{dangling_comments, Comments};
 use crate::expression::parentheses::{
     default_expression_needs_parentheses, NeedsParentheses, Parentheses, Parenthesize,
 };
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
-
-use crate::comments::Comments;
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::trivia::Token;
+use crate::trivia::{first_non_trivia_token, TokenKind};
+use crate::{AsFormat, FormatNodeRule, PyFormatter};
+use ruff_formatter::prelude::{hard_line_break, space, text};
+use ruff_formatter::{Format, FormatError, FormatResult};
+use ruff_python_ast::node::AstNode;
+use ruff_python_ast::prelude::{Expr, Ranged};
+use ruff_text_size::TextRange;
 use rustpython_parser::ast::ExprSlice;
 
 #[derive(Default)]
 pub struct FormatExprSlice;
 
 impl FormatNodeRule<ExprSlice> for FormatExprSlice {
-    fn fmt_fields(&self, _item: &ExprSlice, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(
-            f,
-            [not_yet_implemented_custom_text(
-                "NOT_IMPLEMENTED_start:NOT_IMPLEMENTED_end"
-            )]
-        )
+    /// This implementation deviates from black in that comments are attached to the section of the
+    /// slice they originate in
+    fn fmt_fields(&self, item: &ExprSlice, f: &mut PyFormatter) -> FormatResult<()> {
+        // `[lower:upper:step]`
+        let ExprSlice {
+            range,
+            lower,
+            upper,
+            step,
+        } = item;
+
+        let (first_colon, second_colon) =
+            find_colons(f.context().contents(), *range, lower, upper)?;
+
+        // Handle comment placement
+        // In placements.rs, we marked comment for None nodes a dangling and associated all others
+        // as leading or dangling wrt to a node. That means we either format a node and only have
+        // to handle newlines and spacing, or the node is None and we insert the corresponding
+        // slice of dangling comments
+        let comments = f.context().comments().clone();
+        let slice_dangling_comments = comments.dangling_comments(item.as_any_node_ref());
+        // Put the dangling comments (where the nodes are missing) into buckets
+        let first_colon_partition_index = slice_dangling_comments
+            .partition_point(|x| x.slice().start() < first_colon.range.start());
+        let second_colon_partition_index =
+            second_colon
+                .as_ref()
+                .map_or(slice_dangling_comments.len(), |second_colon| {
+                    slice_dangling_comments
+                        .partition_point(|x| x.slice().start() < second_colon.range.start())
+                });
```

Comment:
Nit: Can we optimize this to only iterate over the *after first colon* comments

```suggestion
		let (first_colon_comments, remaining_comments) = slice_dangling_comments.split_at(first_colon_partition_index);
		
        let second_colon_partition_index = 
            second_colon
                .as_ref()
                .map_or(remaining_comments.len(), |second_colon| {
                    remaining_comments
                        .partition_point(|x| x.slice().start() < second_colon.range.start())
                });
```

Reply:
not with that snippet, but a different solution works